### PR TITLE
[validation] Improve error message for unresolved methods. Bug 377775

### PR DESCRIPTION
**Pull request by dhuebner**
Signed-off-by: Dennis Huebner <dennis.huebner@itemis.de>

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=377775

**Comment by svenefftinge** [comment]
Could we move the implementation to xbase?

**Comment by dhuebner** [comment]
I need to solve the LinkingErrorTest problem first :(

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/IssueCodes.java [hunk 1]
```diff
@@ -17,7 +17,7 @@ private IssueCodes() {
 	}
 
 	protected static final String ISSUE_CODE_PREFIX = "org.eclipse.xtend.core.validation.IssueCodes.";
-	public static final String FEATURECALL_LINKING_DIAGNOSTIC = ISSUE_CODE_PREFIX + "feature_linking";
+//	public static final String FEATURECALL_LINKING_DIAGNOSTIC = ISSUE_CODE_PREFIX + "feature_linking";
```
Comment (dhuebner) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/IssueCodes.java]: Will remove commented code

**Comment by dhuebner** [comment]
Jenkins xtend-ide-tests-master-latest

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/IssueCodes.java (hunk 1 above)
Comment (szarnekow) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/validation/IssueCodes.java]: That'll be a new commit?

File: plugins/org.eclipse.xtext.junit4/src/org/eclipse/xtext/junit4/validation/ValidationTestHelper.java [hunk 2]
```diff
@@ -183,11 +191,21 @@ public boolean apply(Issue input) {
...
+	 * @since 2.9
+	 */
+	public void assertNoError(final Resource resource, final String issuecode, final String userData) {
 		final List<Issue> validate = validate(resource);
 		Iterable<Issue> issues = filter(validate, new Predicate<Issue>() {
 			@Override
 			public boolean apply(Issue input) {
-				return issuecode.equals(input.getCode());
+				if (issuecode.equals(input.getCode())) {
+					return userData == null || Iterables.contains(Arrays.asList(input.getData()), userData);
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.junit4/src/org/eclipse/xtext/junit4/validation/ValidationTestHelper.java]: Minor: Arrays.asList(..).contains(..) would be slightly more readable.

File: plugins/org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/annotations/validation/LinkingDiagnosticTypeAwareMessageProducer.xtend [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import org.eclipse.xtext.xbase.util.FeatureCallAsTypeLiteralHelper
+
+import static org.eclipse.xtext.diagnostics.Diagnostic.LINKING_DIAGNOSTIC
+import static org.eclipse.xtext.diagnostics.Severity.ERROR
+import static org.eclipse.xtext.xbase.XbasePackage.Literals.XMEMBER_FEATURE_CALL__MEMBER_CALL_TARGET
+
+/** 
+ * @author Dennis Huebner - Initial contribution and API
+ */
+class LinkingDiagnosticTypeAwareMessageProducer extends LinkingDiagnosticMessageProvider {
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/annotations/validation/LinkingDiagnosticTypeAwareMessageProducer.xtend]: This class name reads a little be weird. Can this be TypeAwareLinkingDiagnosticMessageProducer ?

File: tests/org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/FeatureCallValidationTest.java [hunk 4]
```diff
@@ -255,6 +256,53 @@ public void testBug400989() throws Exception {
...
+	@Test
+	public void testBug334499_05() throws Exception {
+		XtendClass clazz = clazz("class C { def void m() { this.foo(this) }}");
+		helper.assertError(clazz, XbasePackage.Literals.XMEMBER_FEATURE_CALL, Diagnostic.LINKING_DIAGNOSTIC, "foo(C) is", "for the type C");
+	}
+	
+	@Test
+	public void testBug334499_06() throws Exception {
+		XtendClass clazz = clazz("class C { def m() { new C(){ override m(){ foo(1)} }}}");
+		helper.assertError(clazz, XbasePackage.Literals.XFEATURE_CALL, Diagnostic.LINKING_DIAGNOSTIC, "foo(int) is");
```
Comment (szarnekow) [inline tests/org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/validation/FeatureCallValidationTest.java]: Could you please add a test for an unresolved call in an anonymous class with a this. receiver?

File: tests/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/QuickfixTestBuilder.xtend [hunk 5]
```diff
@@ -80,6 +82,22 @@ class QuickfixTestBuilder {
 		this
 	}
 	
+	def assertFatureCallLinkingIssue() {
```
Comment (szarnekow) [inline tests/org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/QuickfixTestBuilder.xtend]: Typo: assertF_e_atureCallLinkingIssue

**Comment by szarnekow** [comment]
I like the improved error messages!

File: plugins/org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/annotations/validation/LinkingDiagnosticTypeAwareMessageProducer.xtend (hunk 3 above)
Comment (dhuebner) [inline plugins/org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/annotations/validation/LinkingDiagnosticTypeAwareMessageProducer.xtend]: Inspired by org.eclipse.xtext.xbase.annotations.validation.UnresolvedAnnotationTypeAwareMessageProducer

**Comment by dhuebner** [comment]
I will create a new PR for the suggestions provided here.
https://github.com/eclipse/xtext/pull/469

**Pull request merged**